docs/examples/Fine-Tuning-Function-Calling.md

```diff
@@ -0,0 +1,273 @@
+Here is the command to run the training job on TPU:
+
+```bash
+gcloud compute tpus tpu-vm ssh finetune-32 --zone us-east1-d --worker=all \
+--command="WANDB_API_KEY=${YOUR WANDB TOKEN HERE} \
+HUGGING_FACE_HUB_TOKEN=${YOUR HF TOKEN HERE} \
+bash levanter/infra/run.sh python \
+levanter/examples/alpaca/alpaca.py \
+--config_path gs://<config-yaml-file> \
+--hf_save_path gs://<somewhere>"
+```
+
+Given the small dataset and high efficiency of Levanter, the entire training job completed quickly in only 21 min on a single TPUv3-8.
+
+Below shows an example where we compare the output from the original chat model and the fine-tuned model. We can see that the fine-tuned model is able to generate the correct function call and precisely capture the right attributes, while the original model is not able to do so.
+
+```
+Query: SpellForce 3 is a pretty bad game. The developer Grimlore Games is clearly a bunch of no-talent hacks, and 2017 was a terrible year for games anyway.
+Target: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
+Generation from Llama2-7B Chat: \nOutput: inform(name[SpellForce 3], release_year[2017], esrb[NA], genres[fantasy,\naction, adventure], platforms[PC], available_on_steam[no], has_linux_release[no],\nhas_mac_release[no])
+Generation from fine-tuned Llama2-7B: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
+```
+
+#### LoRA Fine-tuning
+
+Below is our configuration for LoRA fine-tuning. Note that it is very similar to the full-weight fine-tuning configuration, except for a few differences:
+
+- We increased the number of steps by 1 more epoch. LoRA fine-tuning is more efficient and less prone to overfitting, so we can train for more steps.
+- We increased the learning rate to 3e-4, but we did not do very thorough hyperparameter tuning. We expect there might be a better learning rate.
+- We found weight decay at 0.1 to lead to better results than no weight decay, so we set it at 0.1.
+- We added the `lora` section to specify the LoRA parameters. All of the parameters are set to the default values.
+
+```yaml
+data: "gs://levanter-data/fine-tuning/gem-viggo/GEM_viggo_train.jsonl"
+data_cache_dir: "gs://levanter-data/tokenized/GEM_viggo_lora/"
+trainer:
+  wandb:
+    project: "levanter"
+    tags: ["viggo", "llama", "lora"]
+
+  mp: p=f32,c=bfloat16
+  train_batch_size: 128
+  num_train_steps: 120  # 5103 examples / 128 batch size * 3 epochs
+  tensor_parallel_axes: ["mlp", "heads"]
+optimizer:
+  learning_rate: 3e-4
+  weight_decay: 0.1
+lora:
+  r: 8  # rank of LoRA transform
+  alpha: 8.0  # scaling factor for LoRA transform
+  dropout: 0.0  # dropout probability for LoRA layers
+```
+
+After training, Levanter will automatically merge the new parameters into the original model and save the new model. We can use it for inference and evaluation.
```

Comment:
this depends on how it's invoked. with `hf_save_path` it will just save the lora adapter that works with PEFT (so just the lora weights plus a pointer), with merged_hf_save_path it will save a merged checkpoint.

Reply:
it would be good to show the invocation for alpaca_lora because the new lora tutorial doesn't call alpaca_lora